pkg/apis/core/validation/conditional_validation.go

```diff
@@ -17,24 +17,80 @@ limitations under the License.
 package validation
 
 import (
+	"fmt"
+	"net"
+	"strings"
+
 	"k8s.io/apimachinery/pkg/util/validation/field"
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	api "k8s.io/kubernetes/pkg/apis/core"
 	"k8s.io/kubernetes/pkg/features"
+	netutil "k8s.io/utils/net"
 )
 
 // ValidateConditionalService validates conditionally valid fields.
-func ValidateConditionalService(service, oldService *api.Service) field.ErrorList {
+func ValidateConditionalService(service, oldService *api.Service, allowedIPFamilies []api.IPFamily) field.ErrorList {
 	var errs field.ErrorList
 	// If the SCTPSupport feature is disabled, and the old object isn't using the SCTP feature, prevent the new object from using it
 	if !utilfeature.DefaultFeatureGate.Enabled(features.SCTPSupport) && len(serviceSCTPFields(oldService)) == 0 {
 		for _, f := range serviceSCTPFields(service) {
 			errs = append(errs, field.NotSupported(f, api.ProtocolSCTP, []string{string(api.ProtocolTCP), string(api.ProtocolUDP)}))
 		}
 	}
+
+	if len(allowedIPFamilies) > 0 {
+		defaultFamily := allowedIPFamilies[0]
+		desiredIP := net.ParseIP(service.Spec.ClusterIP)
+
+		desiredFamily := defaultFamily
+		var defaultedFromIP bool
+		switch {
+		case service.Spec.IPFamily != nil:
+			desiredFamily = *service.Spec.IPFamily
+		case netutil.IsIPv6(desiredIP):
+			defaultedFromIP = true
+			desiredFamily = api.IPv6Protocol
+		case desiredIP != nil:
+			defaultedFromIP = true
+			desiredFamily = api.IPv4Protocol
+		}
+
+		if !hasIPFamily(allowedIPFamilies, desiredFamily) {
+			if defaultedFromIP {
+				errs = append(errs, field.Invalid(field.NewPath("service", "spec", "clusterIP"), service.Spec.ClusterIP, fmt.Sprintf("only the following IP address types are allowed: %s", joinIPFamilies(allowedIPFamilies, ", "))))
+			} else {
+				errs = append(errs, field.Invalid(field.NewPath("service", "spec", "ipFamily"), desiredFamily, fmt.Sprintf("only the following families are allowed: %s", joinIPFamilies(allowedIPFamilies, ", "))))
```

Comment:
The logic here is confusing and there are no unit tests of this specifically and I can't tell if it effectively gets tested by any of the later tests...
(The logic is right AFAICT but it's confusing.)

Reply:
Ok, will add tests.